crates/miden-block-prover/src/local_block_prover.rs

```diff
@@ -0,0 +1,302 @@
+    }
+
+    /// Proves the provided [`ProposedBlock`] into a [`ProvenBlock`], **without verifying batches
+    /// and proving the block**.
+    ///
+    /// This is exposed for testing purposes.
+    #[cfg(any(feature = "testing", test))]
+    pub fn prove_without_verification(
+        &self,
+        proposed_block: ProposedBlock,
+    ) -> Result<ProvenBlock, ProvenBlockError> {
+        self.prove_without_verification_inner(proposed_block)
+    }
+
+    /// Proves the provided [`ProposedBlock`] into a [`ProvenBlock`].
+    ///
+    /// The assumptions of this method are that the checks made by construction of a
+    /// [`ProposedBlock`] are enforced.
+    ///
+    /// See [`Self::prove`] for more details.
+    fn prove_without_verification_inner(
+        &self,
+        proposed_block: ProposedBlock,
+    ) -> Result<ProvenBlock, ProvenBlockError> {
+        // Get the block number and timestamp of the new block and compute the tx commitment.
+        // --------------------------------------------------------------------------------------------
+
+        let block_num = proposed_block.block_num();
+        let timestamp = proposed_block.timestamp();
+        let tx_hash = BlockHeader::compute_tx_commitment(proposed_block.affected_accounts());
+
+        // Split the proposed block into its parts.
+        // --------------------------------------------------------------------------------------------
+
+        let (
+            _batches,
+            mut account_updated_witnesses,
+            output_note_batches,
+            created_nullifiers,
+            chain_mmr,
+            prev_block_header,
+        ) = proposed_block.into_parts();
+
+        let prev_block_commitment = prev_block_header.hash();
+
+        // Compute the root of the block note tree.
+        // --------------------------------------------------------------------------------------------
+
+        let note_tree = compute_block_note_tree(&output_note_batches);
+        let note_root = note_tree.root();
+
+        // Insert the created nullifiers into the nullifier tree to compute its new root.
+        // --------------------------------------------------------------------------------------------
+
+        let (created_nullifiers, new_nullifier_root) =
+            compute_nullifiers(created_nullifiers, &prev_block_header, block_num)?;
+
+        // Insert the state commitments of updated accounts into the account tree to compute its new
+        // root.
+        // --------------------------------------------------------------------------------------------
+
+        let new_account_root =
+            compute_account_root(&mut account_updated_witnesses, &prev_block_header)?;
+
+        // Insert the previous block header into the block chain MMR to get the new chain root.
+        // --------------------------------------------------------------------------------------------
+
+        let new_chain_root = compute_chain_root(chain_mmr, prev_block_header);
```

Comment:
A lot of these feel like they would benefit from new types with constructors e.g.

```rust
struct AccountCommitment(Digest);

impl AccountCommitment {
    fn compute(..., ) -> Result<Self>;
}
```

but I think this won't work with all of them given how we've split the types. Though I think maybe I just really dislike free functions :)

Reply:
I think the newtypes are a good idea to look into. We'd have to use them pervasively for it to make sense, so `AccountCommitment`, `ChainCommitment`, `NullifierCommitment`, etc. This feels a bit related to #1069.

For the computation, I like having all block-related logic in this file. I don't like free functions when they're public, but as a helper function it seems fine. If we have `AccountCommitment::compute` and friends, then a lot of the logic here would be split out into other places, which has its own downsides, but it would probably feel less bulky here.